src/b-splines/cubic.jl

```diff
@@ -24,145 +24,56 @@ When we derive boundary conditions we will use derivatives `y_0'(x)` and
-    symp, sympp = Symbol("cp_",d) ,Symbol("cpp_",d)
-    symfx = Symbol("fx_",d)
-    symfx_cub = Symbol("fx_cub_", d)
-    sym_1m_fx_cub = Symbol("one_m_fx_cub_", d)
-    quote
-        $symfx_cub = cub($symfx)
-        $sym_1m_fx_cub = cub(1-$symfx)
-        $symm = SimpleRatio(1,6)*$sym_1m_fx_cub
-        $sym  = SimpleRatio(2,3) - sqr($symfx) + SimpleRatio(1,2)*$symfx_cub
-        $symp = SimpleRatio(2,3) - sqr(1-$symfx) + SimpleRatio(1,2)*$sym_1m_fx_cub
-        $sympp = SimpleRatio(1,6)*$symfx_cub
-    end
+expand_index(::Cubic{BC}, xi::Number, ax::AbstractUnitRange, δx) where BC = (xi-1, xi, xi+1, xi+2)
+expand_index(::Cubic{Periodic}, xi::Number, ax::AbstractUnitRange, δx) =
+    (modrange(xi-1, ax), modrange(xi, ax), modrange(xi+1, ax), modrange(xi+2, ax))
+
+# expand_coefs(::Type{BSpline{Cubic{BC}}}, δx) = cvcoefs(δx)
+# expand_coefs(::Type{BSpline{Cubic{BC}}}, dref, d, δx) = ifelse(d==dref, cgcoefs(δx), cvcoefs(δx))
+# function expand_coefs(::Type{BSpline{Cubic{BC}}}, dref1, dref2, d, δx)
+#     if dref1 == dref2
+#         d == dref1 ? chcoefs(δx) : cvcoefs(δx)
+#     else
+#         d == dref1 | d == dref2 ? cgcoefs(δx) : cvcoefs(δx)
+#     end
+# end
+
+function value_weights(::Cubic, δx)
+    x3, xcomp3 = cub(δx), cub(1-δx)
+    (SimpleRatio(1,6) * xcomp3,
+     SimpleRatio(2,3) - sqr(δx) + SimpleRatio(1,2)*x3,
+     SimpleRatio(2,3) - sqr(1-δx) + SimpleRatio(1,2)*xcomp3,
+     SimpleRatio(1,6) * x3)
 end
 
-"""
-In `gradient_coefficients` for a cubic b-spline we assume that `fx_d = x-ix_d`
-and we define `cX_d` for `X ⋹ {m, _, p, pp}` such that
-
-    cm_d  = p'(fx_d)
-    c_d   = q'(fx_d)
-    cp_d  = q'(1-fx_d)
-    cpp_d = p'(1-fx_d)
-
-where `p` and `q` are defined in the docstring entry for `Cubic`, and
-`fx_d` in the docstring entry for `define_indices_d`.
-"""
-function gradient_coefficients(::Type{BSpline{C}}, d) where C<:Cubic
-    symm, sym, symp, sympp = Symbol("cm_",d), Symbol("c_",d), Symbol("cp_",d), Symbol("cpp_",d)
-    symfx = Symbol("fx_",d)
-    symfx_sqr = Symbol("fx_sqr_", d)
-    sym_1m_fx_sqr = Symbol("one_m_fx_sqr_", d)
-    quote
-        $symfx_sqr = sqr($symfx)
-        $sym_1m_fx_sqr = sqr(1 - $symfx)
-
-        $symm  = -SimpleRatio(1,2) * $sym_1m_fx_sqr
-        $sym   =  SimpleRatio(3,2) * $symfx_sqr     - 2 * $symfx
-        $symp  = -SimpleRatio(3,2) * $sym_1m_fx_sqr + 2 * (1 - $symfx)
-        $sympp =  SimpleRatio(1,2) * $symfx_sqr
-    end
+function gradient_weights(::Cubic, δx)
+    x2, xcomp2 = sqr(δx), sqr(1-δx)
+    (-SimpleRatio(1,2) * xcomp2,
+     -2*δx + SimpleRatio(3,2)*x2,
+     +2*(1-δx) - SimpleRatio(3,2)*xcomp2,
+     SimpleRatio(1,2) * x2)
 end
 
-"""
-In `hessian_coefficients` for a cubic b-spline we assume that `fx_d = x-ix_d`
-and we define `cX_d` for `X ⋹ {m, _, p, pp}` such that
-
-    cm_d  = p''(fx_d)
-    c_d   = q''(fx_d)
-    cp_d  = q''(1-fx_d)
-    cpp_d = p''(1-fx_d)
+hessian_weights(::Cubic, δx) = (1-δx, 3*δx-2, 3*(1-δx)-2, δx)
```

Comment:
I think there might be bugs here (and we know from #122 and #181 that there are bugs in the Hessian coefficients as they stood before). These _should_ just be the derivatives of the gradient coefficients, but as far as I can tell they aren't here: the third weight expression should be `3δx - 5`.

As we don't have any tests for Hessians this is difficult to verify, so leave that change out of this PR if you wish. But it's _really_ nice that the mathematics is so much easier to check here, because the code is much closer to the results of the derivation. Well done!

Reply:
I think I'm going to add tests for Hessians and then insert deliberate `throw`s into code-paths that don't work, just to prevent users from relying on stuff that's broken. The error message will include a plea for help :smile:.

I've noticed many of our tests follow a general pattern, so I'm going to rework our tests so that there's a common core of test utilities. That should make it much easier to add the Hessian tests. Probably the very next priority.

Reply:
I may also ultimately fix the Hessians myself, but once the rewrite is done I have higher priorities (other packages to upgrade to 1.0) since I don't use the hessian code myself.